### DISCUSSION TITLE
Bash Installer Alternatives

### DISCUSSION THREAD

**Discussion by bsrz** [Homebrew/discussions#1382] [Getting started]
Hello 👋 

Given the recent CodeCov breach (https://about.codecov.io/security-update/) which compromized 29'000 customer's code bases, are there any plan on moving to a different style of installation?
Basically `curl | bash` is just dangerous on many levels. Homebrew's installation instructions are using `curl | bash` as default, I'm not sure there are other comprehensive installation options?

**Comment by carlocab** [comment]
There already is an alternative: https://docs.brew.sh/Installation#untar-anywhere

However, AFAICT, alternative installation methods aren't really any more secure than the default installer under most reasonable threat models. (Not an expert, could easily be mistaken here. 🙂)

**Comment by SMillerDev** [comment]
You mean like https://docs.brew.sh/Installation#alternative installs?

**Comment by MikeMcQuaid** [comment]
> are there any plan on moving to a different style of installation?

No.

**Comment by MikeMcQuaid** [comment]
> Basically `curl | bash` is just dangerous on many levels

@bsarrazin Also, it may be worth explaining why you consider this to be dangerous in our specific case i.e. `curl | bash` with the tip of a branch on a tightly controlled repository. What's the threat model you see here and why doesn't it apply to other installation methods?

**Comment by carlocab** [comment]
Relevant discussion here: https://hackerone.com/reports/1166535

**Reply by sjackman** [reply]
@bsarrazin Note that the CodeCov installation script `bash <(curl -s https://codecov.io/bash)` downloads the script from `codecov.io`, whereas the Homebrew installation script `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` downloads the installation script directly from GitHub.

The commits of the `Homebrew/install` git repo are signed using GPG. You can verify the signature of the `HEAD` commit of the `Homebrew/install` GitHub repo is signed, and thus the installation script, using either the GitHub web interface https://github.com/Homebrew/install/commits/HEAD (look for the green `Verified`) or using Git at the command line. If you clone the `Homebrew/install` Git repo, you can verify the signature before you run the `install.sh` script from that same Git repo.

```console
$ gpg --recv-keys 4AEE18F83AFDEB23
gpg: requesting key 3AFDEB23 from hkp server keys.gnupg.net
gpg: key 3AFDEB23: public key "GitHub (web-flow commit signing) <noreply@github.com>" imported
gpg: no ultimately trusted keys found
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
$ git clone --depth=1 https://github.com/Homebrew/install
Cloning into 'install'...
…
$ git -C install verify-commit HEAD
gpg: Signature made Wed 14 Apr 00:41:06 2021 PDT
gpg:                using RSA key 4AEE18F83AFDEB23
gpg: Good signature from "GitHub (web-flow commit signing) <noreply@github.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 5DE3 E050 9C47 EA3C F04A  42D3 4AEE 18F8 3AFD EB23
$ /bin/bash install/install.sh
==> Checking for `sudo` access (which may request your password).
Password:
==> This script will install:
/usr/local/bin/brew
…
```

To suppress the warning
```
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
```
you would need to sign GitHub's public GPG key `4AEE18F83AFDEB23` using your own private GPG to indicate that you trust GitHub key's. See https://serverfault.com/questions/569911/how-to-verify-an-imported-gpg-key

**Reply by jonathancross** [reply]
Thanks @sjackman, this is more secure than the `curl | bash` method. Unfortunatly it is still trusting Github to be correct.  It would be better if Homebrew developers signed the commits or created a detached gpg signature for `install.sh`.

**Reply by carlocab** [reply]
If someone compromised GitHub to get to the Homebrew installer, I doubt extra/alternative signatures is going to help very much. What is the threat model where this provides more security than what is currently done?

**Reply by jonathancross** [reply]
Developers creating GPG signatures with a key they alone control means zero trust in the github platform, their control of their keys, rogue employees, etc.  The simplicity of zero risk in this regard with little overhead is just a step in the right direction, even if it is highly unlikely.  I could imagine a situation where high value targets (Edward Snowden, etc) would be deliberately sent a malicious `install.sh`.   Microsoft (enthusiastic early volunteer for PRISM) owns GitHub -- maybe a situation will eventually arise where corporate / government interests trump reputation and they think they can get away with such a switch.

This install file is a significant, potential single point of failure for thousands (millions?) of users and the foundation of all installs that come afterwards.  It makes sense to give power users a way to do a _tinfoil hat_ install.

Providing a gpg sig from a key you control is defense-in-depth.

**Reply by SMillerDev** [reply]
So how would people get the key? Through our GitHub hosted website?

**Reply by carlocab** [reply]
If Microsoft wanted to compromise the Homebrew installer, it would probably not be too difficult for them to also compromise alternative delivery channels for the signatures, especially since users interact with Homebrew almost exclusively via GitHub. That's not really a threat model that demonstrates that your suggestion would make a meaningful difference to security.

**Reply by jonathancross** [reply]
Ideally the release signing key itself should be certified (via [C] digital signature -- aka "key signing" in pgp lingo) by multiple devs and people in the [OpenPGP Web Of Trust strong set](https://en.wikipedia.org/wiki/Web_of_trust#Strong_set), so that users could be sure it is the correct signing key without trusing centralized authorities or hosting infrastructure.  This is how things are done in security conscious projects like  Bitcoin, Monero, TailsOS, Debian, Tor, etc.

> compromise alternative delivery channels for the signatures

The beauty of this system is that the delivery channels for the signed data AND the signatures no longer need to be trusted.  The data is end-to-end verifiable.
The fingerprint of the file establishes that it has not been modified and the gpg signature establishes what the _correct_ fingerprint should be (according to developers).

**Reply by jonathancross** [reply]
> So how would people get the key? Through our GitHub hosted website?

The public key can be spammed all over the place on your website, github, etc... the key fingerprint should be added to presentations, business cards, etc.  There is keyserver infrastructure[1] which hosts pgp keys for users and makes it easy to find the keys.  The "correct" key is established by the user (decentralized) based on what info they have available such as WOT (see above), checking the website, twitter, watching which key is signing git commits, asking experts, meeting people in person (this is the best, but not necessarily scalable)

[1] https://keyserver.ubuntu.com
https://keys.openpgp.org

**Reply by jonathancross** [reply]
Example:  My key ID is [c0c076132ffa7695](https://keyserver.ubuntu.com/pks/lookup?op=vindex&search=0xc0c076132ffa7695).
I have downloaded the Tails OS from https://tails.boum.org/install/download/ and would like to establish if it is authentic before installing.
1. I get the gpg signature from the download page above.
2. I get the gpg key ([dbb802b258acd84f](https://keyserver.ubuntu.com/pks/lookup?op=vindex&search=0xdbb802b258acd84f)) from the website, a keyserver, etc.
3. I want to know if the key is authentic.
A google search returns [this](https://tor.stackexchange.com/questions/19487/help-with-tails-signing-key-verification), [this](https://www.reddit.com/r/tails/comments/32vlnj/tails_signing_key/), [this](https://www.reddit.com/r/tails/comments/3am0od/how_can_i_tell_tailsboumorg_is_legit/) and [someone I know and trust](https://twitter.com/HillebrandMax/status/1154256823127629824). -- so there is a good chance it is okay.

But I can do better.

I also met these guys at a conferences and verified their keys in person [7fab114267e4fa04](https://keyserver.ubuntu.com/pks/lookup?op=vindex&search=0x7fab114267e4fa04), [1d84ccf010cc5bc7](https://keyserver.ubuntu.com/pks/lookup?op=vindex&search=0x1d84ccf010cc5bc7) -- both are extremely thorough and one is the release engineer... they have both have cryptographically signed the alleged TailsOS signing key and my gpg software automatically calculates the trust level and gives me a nice green set of bars.
![Screen Shot 2021-11-20 at 4 22 21 PM](https://user-images.githubusercontent.com/5115470/142731858-925bd88e-4ec8-4eb6-816e-7ba3b13dff3c.png)

Eventually, I went to a conference where TailsOS had a booth and they had the key fingerprint on a big banner and on business cards.
I therefore "signed" their key (technically a certification signature) with mine.

I can now be extremely confident that the tails I downloaded from bittorrent, the website or wherever is the one that the devs created:
```
gpg --verify tails-amd64-4.24.img.sig
gpg: assuming signed data in 'tails-amd64-4.24.img'
gpg: Signature made Thu Nov  4 15:06:02 2021 CET
gpg:                using RSA key 05469FB85EAD6589B43D41D3D21DAD38AF281C0B
gpg: Good signature from "Tails developers (offline long-term identity key) <tails@boum.org>" [full]
gpg:                 aka "Tails developers <tails@boum.org>" [full]
Primary key fingerprint: A490 D0F4 D311 A415 3E2B  B7CA DBB8 02B2 58AC D84F
     Subkey fingerprint: 0546 9FB8 5EAD 6589 B43D  41D3 D21D AD38 AF28 1C0B
```


Note: I am using shorter key ID's to cut down on noise, but full SHA1 fingerprint should be used when verifying.

**Comment by MikeMcQuaid** [comment]
> Unfortunatly it is still trusting Github to be correct

Yes, this is a decision we've made.

> Microsoft (enthusiastic early volunteer for PRISM) owns GitHub -- maybe a situation will eventually arise where corporate / government interests trump reputation and they think they can get away with such a switch.

More easily: government interests could compel a Homebrew maintainer to sign a "bad" release.

At the point that you're worrying about government-level attackers as a volunteer run project that's not solely a security tool: you're just playing a game.

> Ideally the release signing key itself should be certified (via [C] digital signature -- aka "key signing" in pgp lingo) by multiple devs and people in the [OpenPGP Web Of Trust strong set](https://en.wikipedia.org/wiki/Web_of_trust#Strong_set)

We're not going to do this.

---

You're welcome to send PRs to improve the security profile of Homebrew. We're volunteers here and we're not going to sign up for more work for minimal gain. We have had (and it's on my TODO to write up) a proper, 3rd-party security audit funded by Mozilla and they didn't make this recommendation.

**Reply by SMillerDev** [reply]
Often people forget the easiest scenario:
![xkcd: security](https://imgs.xkcd.com/comics/security.png)

**Reply by MikeMcQuaid** [reply]
> We have had (and it's on my TODO to write up)

It got wrote up: https://brew.sh/2022/05/17/homebrew-security-audit/

**Comment by gwerbin** [comment]
Sorry to bump an old-ish thread, but what about providing an SHA256 checksum for the installer script? I've had a couple of times where a download simply failed for some reason or other, and left whatever I was downloading in some kind of mangled state.

Of course the `git clone` solution offered by sjackman will take care of that, but it's not an officially supported option in the docs.

**Reply by MikeMcQuaid** [reply]
Yeh, we don't want to support more installation options, sorry, and the checksum wouldn't really solve anything a `git clone` will.

**Reply by gwerbin** [reply]
@MikeMcQuaid but is the `git clone` solution supported officially or not? If not, then there's no official way to check that the download succeeded until you run the script and verify that it worked.

**Reply by MikeMcQuaid** [reply]
> then there's no official way to check that the download succeeded until you run the script and verify that it worked.

yes
